File: templates/problems/13_dp_1d/10_word_break/go/solution.go
package word_break

// solveWordBreak checks if s can be segmented into dictionary words. O(n^2) time, O(n) space.
func solveWordBreak(s string, wordDict []string) bool {
	wordSet := make(map[string]bool, len(wordDict))
	for _, w := range wordDict {
		wordSet[w] = true
	}
	n := len(s)
	dp := make([]bool, n+1)
	dp[0] = true
	for i := 1; i <= n; i++ {
		for j := 0; j < i; j++ {
			if dp[j] && wordSet[s[j:i]] {
				dp[i] = true
				break
			}
		}
	}
	return dp[n]
}

File: templates/problems/13_dp_1d/10_word_break/python/solution.py
def solve_word_break(s: str, word_dict: list[str]) -> bool:
    """Check if s can be segmented into dictionary words.

    Time: O(n^2), Space: O(n)
    """
    word_set = set(word_dict)
    n = len(s)
    dp = [False] * (n + 1)
    dp[0] = True
    for i in range(1, n + 1):
        for j in range(i):
            if dp[j] and s[j:i] in word_set:
                dp[i] = True
                break
    return dp[n]
